doc/man/fltcopy.1
.\" Automatically generated by fltcopy -write-bam
.TH FLTCOPY 1 "27 December 2014" "1.9.0" Panda3D
.SH NAME
fltcopy \- copy MultiGen .flt files into a CVS source hierarchy
.SH SYNOPSIS
\fBfltcopy\fR [opts] file.flt [file.flt ... ]
.SH DESCRIPTION
fltcopy copies one or more MultiGen .flt files into a CVS source hierarchy.  Rather than copying the named files immediately into the current directory, it first scans the entire source hierarchy, identifying all the already\-existing files.  If the named file to copy matches the name of an already\-existing file in the current directory or elsewhere in the hierarchy, that file is overwritten.  Other .flt files, as well as texture files, that are externally referenced by the named .flt file(s) are similarly copied.
.SH OPTIONS
.TP
.BI "\-pr " "path_replace"
Sometimes references to other files (textures, external references) are stored with a full path that is appropriate for some other system, but does not exist here.  This option may be used to specify how those invalid paths map to correct paths.  Generally, this is of the form 'orig_prefix=replacement_prefix', which indicates a particular initial sequence of characters that should be replaced with a new sequence; e.g. '/c/home/models=/beta/fish'.  If the replacement prefix does not begin with a slash, the file will then be searched for along the search path specified by -pp.  You may use standard filename matching characters ('*', '?', etc.) in the original prefix, and '**' as a component by itself stands for any number of components.

This option may be repeated as necessary; each file will be tried against each specified method, in the order in which they appear in the command line, until the file is found.  If the file is not found, the last matching prefix is used anyway.
.TP
.BI "\-pp " "dirname"
Adds the indicated directory name to the list of directories to search for filenames referenced by the source file.  This is used only for relative paths, or for paths that are made relative by a -pr replacement string that doesn't begin with a leading slash.  The model-path is always implicitly searched anyway.
.TP
.B \-f
Force the copy to happen without any input from the user.  If a file with the same name exists anywhere in the source hierarchy, it will be overwritten without prompting; if a file does not yet exist, it will be created in the directory named by -d or by -m, as appropriate.
.TP
.B \-i
The opposite of -f, this will prompt the user before each action.  The default is only to prompt the user when an action is ambiguous or unusual.
.TP
.BI "\-d " "dirname"
Copy model files that are not already present somewhere in the tree to the indicated directory.  The default is the current directory.
.TP
.BI "\-m " "dirname"
Copy texture map files to the indicated directory.  The default is src/maps from the root directory.
.TP
.BI "\-root " "dirname"
Specify the root of the CVS source hierarchy.  The default is to use the ppremake convention of locating the directory above the -d directory that contains a file called Package.pp.
.TP
.BI "\-key " "filename"
Specify the name of the file that must exist in each directory for it to be considered part of the CVS source hierarchy.  The default is the ppremake convention, "Sources.pp".  Other likely candidates are "CVS" to search a CVS hierarchy, or "." to include all subdirectories indiscriminately.
.TP
.B \-nc
Do not attempt to add newly-created files to CVS.  The default is to add them.
.TP
.BI "\-cvs " "cvs_binary"
Specify how to run the cvs program for adding newly-created files.  The default is simply "cvs".
.TP
.B \-h
Display this help page.


doc/ReleaseNotes
------------------------  RELEASE 1.8.1  ------------------------

This is a bugfix release, fixing many issues in 1.8.0.
However, there may still be some (minor) bugs.

* Fix a host of issues related to GLSL shaders
* Fix incorrect registry entry for Python in Windows installer
* pdeploy generated binaries with wrong architecture on Linux
* Fix runtime error in pdeploy when building for Windows
* Support for Maya 2013
* ARToolKit now also works on Mac OS X
* WM_CLASS can be set using x-wm-class and x-wm-class-name
* No longer crashes when Xrandr is not supported
* Aux normals are now also normalized when no lights are applied
* Fix hidden cursor when switching fullscreen on Mac OS X
* PackageInstaller didn't add packages to system paths
* Allow disabling custom cursor on Windows
* Fix incorrect panning of 3D audio
* Fix omission of textures of non-standard format in ShaderGenerator
* Fix compile issue with newer gcc versions
* Fix circular reference held by ActorNode
* Window is now correctly centered on Windows
* Fix confusion with depth range of Lens::project()
* Now successfully compiles against recent SSL versions on Windows
* Fix incorrect Cg TEXUNIT0 binding during the first frame
* TextNode::set_text_scale now correctly scales spaces as well
* Expose AudioLoadRequest to Python (for async audio loading)
* Support for the libRocket debugger
* Fix newline entry in libRocket

------------------------  RELEASE 1.8.0  ------------------------

This is a major release, with several big new features.  As such,
it is likely to contain bugs.

* True threading support now enabled in the default build
* Pipelined rendering: app, cull, and draw can run in parallel, each
  in their own thread
* Web plugin is more robust, and better supports Safari and Chrome
* Plugin runs properly when the username contains non-ASCII characters on Windows
* Added appRunner.p3dFilename and appRunner.p3dUrl to provide p3d
  location
* Multifiles (and p3d files) now make a distinction between binary and
  text files
* OccluderNode added for explicit occlusion culling
* Ambient occlusion generation for terrain
* Fixed bug where Windows installer wipes %PATH% when it's too long
* Added fog support to the shader generator
* Added normal_gloss texture mode
* Added a custom color option to the cartoon filter
* Support for texture arrays in shaders
* Better shader support in pandadx9
* Fix some issues with cube map buffers
* Can be compiled to use double-precision floats throughout, instead
  of the default of single-precision floats.  (Graphics drivers still
  use single-precision floats, of course.)
* Can be compiled with the optional Eigen library to provide SSE2 support
* Can be compiled with SpeedTree support
* Can be compiled on MSVS2010, and/or Win64.  (These builds not
  provided by default.)
* Support for the Bullet physics engine
* Support for the libRocket GUI library
* Support for stereo/multiview textures
* Substantial performance improvements to movie textures
* TGA files with alpha channel now load correctly
* New "Ramdisk" mount type available for the VFS
* The VFS is now writable for ramdisk files and true on-disk files
* pdeploy -i generates a custom icon for the installed game
* wx and tk work better on OSX
* Panda windows can be embedded within wxPython windows on all
  platforms (including OSX) with the new WxPandaWindow class
* Added base.pixel2d for pixel-based 2-D coordinates
* Python-based swizzling of Panda vectors, e.g. vec2.xyxy
* Python programmers can now optionally use the original unmangled C++
  name for methods and classes, e.g. model.set_pos(LPoint3f(1, 2, 3)).
* Command-line filename globbing now supported on Win32, e.g. egg-texture-cards *.png
* DirectGui works with nonstandard coordinate-system in effect
* Egg loader handles double-sided polygons a little differently by
  default now, for better render performance but more memory usage
  (use "egg-emulate-bface 0" to restore the old behavior if needed).

------------------------  RELEASE 1.7.2  ------------------------

This release fixes several bugs that were found in 1.7.1.

* Fix crash on GLX implementations that have no FBConfig support
* Fix trouble with buffers on Mac OS X
* Un-break shadow samplers in Cg shaders
* Fixes for relative mouse mode on OSX
* Pdeployed apps on Windows no longer show a console window
* Fix relative file paths for license files in pdeploy
* Ppatcher no longer writes out faulty checksums
* Fix plugin failure to read from cache
* Include missing X11 extension libs in runtime distribution
* Fix disappearing windows with CEGUI's OpenGL renderer
* Fixes for makepanda on FreeBSD
* Fix LightRampAttrib crash
* Fix bug with two-parameter Lens::set_fov

------------------------  RELEASE 1.7.1  ------------------------

This release introduces several significant bugfixes,
but also introduces various minor new features. Although
it is a minor release, it may also introduce new bugs.

* Many improvements and bugfixes to pdeploy
* Vectors now support swizzle/write masks (e.g vec.xz)
* Fixes for depth buffer instabilities on Windows
* Better webcam support on Linux using Video4Linux
* Custom cursor support in X11
* Static functions that return a list are now properly wrapped
* ODE objects now have getId() exposed to Python
* NodePath.findMaterial now works properly
* Remove unnecessary dependency on GLU
* Arithmetic operators to PNMImage
* Various OpenGL ES-related bugfixes
* Support for EGL and OpenGL ES in makepanda
* Fix a crash with the Maya converters
* Include missing p3d tools on Windows
* Include tinyxml as part of the source
* Updates to PandAI
* Compile issues with latest OpenSSL fixed
* Fix static-init ordering issues with OpenSSL
* Several other bugfixes and features not listed here

------------------------  RELEASE 1.7.0  ------------------------

This major release introduces tons of cool new features. As it
is highly experimental, it is not recommended for production use.


pandatool/src/fltprogs/fltCopy.cxx
// Filename: fltCopy.cxx
// Created by:  drose (01Nov00)
//
////////////////////////////////////////////////////////////////////
//
// PANDA 3D SOFTWARE
// Copyright (c) Carnegie Mellon University.  All rights reserved.
//
// All use of this software is subject to the terms of the revised BSD
// license.  You should have received a copy of this license along
// with this source code in a file named "LICENSE."
//
////////////////////////////////////////////////////////////////////

#include "fltCopy.h"

#include "cvsSourceDirectory.h"
#include "fltHeader.h"
#include "fltFace.h"
#include "fltExternalReference.h"
#include "fltError.h"
#include "dcast.h"
#include "pystub.h"

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::Constructor
//       Access: Public
//  Description:
////////////////////////////////////////////////////////////////////
FltCopy::
FltCopy() {
  set_program_brief("copy MultiGen .flt files into a CVS source hierarchy");
  set_program_description
    ("fltcopy copies one or more MultiGen .flt files into a "
     "CVS source hierarchy.  "
     "Rather than copying the named files immediately into the current "
     "directory, it first scans the entire source hierarchy, identifying all "
     "the already-existing files.  If the named file to copy matches the "
     "name of an already-existing file in the current directory or elsewhere "
     "in the hierarchy, that file is overwritten.  Other .flt files, as "
     "well as texture files, that are externally referenced by the "
     "named .flt file(s) are similarly copied.");

  clear_runlines();
  add_runline("[opts] file.flt [file.flt ... ]");

  add_path_replace_options();
}

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::run
//       Access: Public
//  Description:
////////////////////////////////////////////////////////////////////
void FltCopy::
run() {
  SourceFiles::iterator fi;
  for (fi = _source_files.begin(); fi != _source_files.end(); ++fi) {
    ExtraData ed;
    ed._type = FT_flt;

    CVSSourceTree::FilePath dest = import(*fi, &ed, _model_dir);
    if (!dest.is_valid()) {
      exit(1);
    }
  }
}

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::copy_file
//       Access: Protected, Virtual
//  Description: Called by import() if verify_file() indicates that a
//               file needs to be copied.  This does the actual copy
//               of a file from source to destination.  If new_file is
//               true, then dest does not already exist.
////////////////////////////////////////////////////////////////////
bool FltCopy::
copy_file(const Filename &source, const Filename &dest,
          CVSSourceDirectory *dir, void *extra_data, bool new_file) {
  ExtraData *ed = (ExtraData *)extra_data;
  switch (ed->_type) {
  case FT_flt:
    return copy_flt_file(source, dest, dir);

  case FT_texture:
    return copy_texture(source, dest, dir, ed->_texture, new_file);
  }

  nout << "Internal error: invalid type " << (int)ed->_type << "\n";
  return false;
}

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::copy_flt_file
//       Access: Private
//  Description:
////////////////////////////////////////////////////////////////////
bool FltCopy::
copy_flt_file(const Filename &source, const Filename &dest,
              CVSSourceDirectory *dir) {
  PT(FltHeader) header = new FltHeader(_path_replace);

  // We don't want to automatically generate .attr files--we'd rather
  // write them out explicitly.
  header->set_auto_attr_update(FltHeader::AU_none);

  FltError result = header->read_flt(source);
  if (result != FE_ok) {
    nout << "Cannot read " << source << ": " << result << "\n";
    return false;
  }

  header->check_version();

  // Now scan the flt file for nested references.
  Refs refs;
  Textures textures;
  scan_flt(header, refs, textures);

  Refs::const_iterator ri;
  for (ri = refs.begin(); ri != refs.end(); ++ri) {
    FltExternalReference *ref = (*ri);
    Filename ref_filename = ref->get_ref_filename();

    if (!ref_filename.exists()) {
      nout << "*** Warning: external reference " << ref_filename
           << " does not exist.\n";
    } else {
      ExtraData ed;
      ed._type = FT_flt;

      CVSSourceTree::FilePath ref_path =
        import(ref_filename, &ed, _model_dir);
      if (!ref_path.is_valid()) {
        return false;
      }

      // Update the reference to point to the new flt filename, relative
      // to the base flt file.
      ref->set_ref_filename(ref_path.get_rel_from(dir));
    }
  }

  // Remove all the textures from the palette, and then add back only
  // those we found in use.  This way we don't copy a file that
  // references bogus textures.
  header->clear_textures();

  Textures::const_iterator ti;
  for (ti = textures.begin(); ti != textures.end(); ++ti) {
    FltTexture *tex = (*ti);
    Filename texture_filename = tex->get_texture_filename();

    if (!texture_filename.exists()) {
      nout << "*** Warning: texture " << texture_filename
           << " does not exist.\n";
    } else {
      ExtraData ed;
      ed._type = FT_texture;
      ed._texture = tex;

      CVSSourceTree::FilePath texture_path =
        import(texture_filename, &ed, _map_dir);
      if (!texture_path.is_valid()) {
        return false;
      }

      // Update the texture reference to point to the new texture
      // filename, relative to the flt file.
      tex->set_texture_filename(texture_path.get_rel_from(dir));
      header->add_texture(tex);
    }
  }

  // Finally, write the resulting file out.
  result = header->write_flt(dest);
  if (result != FE_ok) {
    nout << "Cannot write " << dest << "\n";
    return false;
  }

  return true;
}

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::copy_texture
//       Access: Private
//  Description:
////////////////////////////////////////////////////////////////////
bool FltCopy::
copy_texture(const Filename &source, const Filename &dest,
             CVSSourceDirectory *dir, FltTexture *tex, bool new_file) {
  if (!copy_binary_file(source, dest)) {
    return false;
  }

  // Also write out the .attr file.
  Filename attr_filename = dest.get_fullpath() + ".attr";
  if (!attr_filename.exists()) {
    new_file = true;
  }

  tex->write_attr_data(attr_filename);

  if (new_file) {
    cvs_add(attr_filename);
  }

  return true;
}

////////////////////////////////////////////////////////////////////
//     Function: FltCopy::scan_flt
//       Access: Private
//  Description: Recursively walks through the flt file hierarchy,
//               looking for texture references and external flt file
//               references.
////////////////////////////////////////////////////////////////////
void FltCopy::
scan_flt(FltRecord *record, FltCopy::Refs &refs, FltCopy::Textures &textures) {
  if (record->is_of_type(FltFace::get_class_type())) {
    FltFace *face;
    DCAST_INTO_V(face, record);
    if (face->has_texture()) {
      textures.insert(face->get_texture());
    }

  } else if (record->is_of_type(FltExternalReference::get_class_type())) {
    FltExternalReference *ref;
    DCAST_INTO_V(ref, record);

    refs.insert(ref);
  }

  int i;
  int num_subfaces = record->get_num_subfaces();
  for (i = 0; i < num_subfaces; i++) {
    scan_flt(record->get_subface(i), refs, textures);
  }

  int num_children = record->get_num_children();
  for (i = 0; i < num_children; i++) {
    scan_flt(record->get_child(i), refs, textures);
  }
}


int main(int argc, char *argv[]) {
  // A call to pystub() to force libpystub.so to be linked in.
  pystub();

  FltCopy prog;
  prog.parse_command_line(argc, argv);
  prog.run();
  return 0;
}
